File: python/734.py
"""Project Euler Problem 734: A Bit of Prime.

Find the number of K-tuples (x_1, x_2, ... x_K) such that each element is a
prime ≤ N and the bitwise OR of all x_i is also a prime ≤ N.

Let c_i = 1 if i is prime, and 0 otherwise, and let X = Σ_i c_i x^i. If we
define multiplication by (x^i)(x^j) = x^{i|j} (| is the bitwise OR), then we
can see that each coefficient c_i of X^K represents the number of ways that the
bitwise OR of K of these primes equals i. The answer is the sum of all c_p for
p prime ≤ N.

To efficiently compute X^K, we use a variation of nimPow, using a different
Hadamard algorithm that works for bitwise OR instead of bitwise XOR.
"""

from __future__ import annotations

from math import ceil, log2

from sympy import primerange


def iceil_pow(n: int, base: int) -> int:
    """Smallest power of base >= n."""
    if n <= 1:
        return 1
    exp = ceil(log2(n) / log2(base))
    return base**exp


def pow_mod(base: int, exp: int, mod: int) -> int:
    """Modular exponentiation."""
    if mod <= 0:
        msg = "mod must be positive"
        raise ValueError(msg)
    if base == 0:
        return 0
    base %= mod
    result = 1
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result


def solve() -> int:
    """Solve Problem 734."""
    N = 10**6
    K = 999983
    M = 10**9 + 7
    L = iceil_pow(N, 2)

    # Initialize array with primes
    primes_list = list(primerange(2, N + 1))
    A = [0] * L
    for p in primes_list:
        if p < L:
            A[p] = 1

    # Forward Hadamard transform for bitwise OR
    u = 1
    while u < L:
        for x in range(0, L):
            if x & u == 0:  # Only process if bit is not set
                A[x | u] = (A[x | u] + A[x]) % M
        u <<= 1

    # Raise to K-th power
    for i in range(L):
        A[i] = pow_mod(A[i], K, M)

    # Inverse Hadamard transform
    u = 1
    while u < L:
        for x in range(0, L):
            if x & u == 0:
                A[x | u] = (A[x | u] - A[x]) % M
                if A[x | u] < 0:
                    A[x | u] += M
        u <<= 1

    # Sum over primes
    ans = 0
    for p in primes_list:
        if p < L:
            ans = (ans + A[p]) % M

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/734.c
/* Project Euler 734: A Bit of Prime.
 * OR-convolution via Mobius/Zeta transform on bitwise OR.
 * N=10^6, K=999983, M=10^9+7.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef long long ll;
typedef __int128 lll;

#define MOD 1000000007LL

ll pow_mod(ll base, ll exp, ll mod) {
    ll result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp & 1) result = (lll)result * base % mod;
        base = (lll)base * base % mod;
        exp >>= 1;
    }
    return result;
}

int main() {
    int N = 1000000;
    int K = 999983;

    /* Find L = smallest power of 2 >= N */
    int L = 1;
    while (L < N) L <<= 1;

    /* Sieve of Eratosthenes */
    char *is_prime = calloc(N + 1, 1);
    for (int i = 2; i <= N; i++) is_prime[i] = 1;
    for (int i = 2; (ll)i * i <= N; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= N; j += i)
                is_prime[j] = 0;
        }
    }

    /* Initialize array */
    ll *A = calloc(L, sizeof(ll));
    int *primes = NULL;
    int nprimes = 0;
    for (int p = 2; p <= N; p++) {
        if (is_prime[p]) {
            A[p] = 1;
            nprimes++;
        }
    }
    primes = malloc(nprimes * sizeof(int));
    int idx = 0;
    for (int p = 2; p <= N; p++)
        if (is_prime[p]) primes[idx++] = p;

    /* Forward Zeta transform for bitwise OR (subset sum) */
    for (int u = 1; u < L; u <<= 1) {
        for (int x = 0; x < L; x++) {
            if ((x & u) == 0) {
                A[x | u] = (A[x | u] + A[x]) % MOD;
            }
        }
    }

    /* Raise to K-th power */
    for (int i = 0; i < L; i++) {
        A[i] = pow_mod(A[i], K, MOD);
    }

    /* Inverse Mobius transform */
    for (int u = 1; u < L; u <<= 1) {
        for (int x = 0; x < L; x++) {
            if ((x & u) == 0) {
                A[x | u] = (A[x | u] - A[x] % MOD + MOD) % MOD;
            }
        }
    }

    /* Sum over primes */
    ll ans = 0;
    for (int i = 0; i < nprimes; i++) {
        int p = primes[i];
        if (p < L)
            ans = (ans + A[p]) % MOD;
    }

    printf("%lld\n", ans);

    free(A);
    free(is_prime);
    free(primes);
    return 0;
}
